TOOLS/lua/autocrop.lua

```diff
@@ -1,114 +1,341 @@
-                                     cropdetect_metadata["lavfi.cropdetect.h"],
-                                     cropdetect_metadata["lavfi.cropdetect.x"],
-                                     cropdetect_metadata["lavfi.cropdetect.y"]))
-        else
-            mp.msg.error(
-                "Got empty crop data. You might need to increase detect_seconds."
-            )
-        end
+        meta = {
+            w = cropdetect_metadata["lavfi.cropdetect.w"],
+            h = cropdetect_metadata["lavfi.cropdetect.h"],
+            x = cropdetect_metadata["lavfi.cropdetect.x"],
+            y = cropdetect_metadata["lavfi.cropdetect.y"],
+        }
     else
-        mp.msg.error(
-            "No crop data. Was the cropdetect filter successfully inserted?"
+        mp.msg.error("No crop data.")
+        mp.msg.info("Was the cropdetect filter successfully inserted?")
+        mp.msg.info("Does your version of ffmpeg/libav support AVFrame metadata?")
+        return
+    end
+
+    -- Verify that the metadata meets the requirements and convert it.
+    if meta.w and meta.h and meta.x and meta.y then
+        meta = {
+            w = tonumber(meta.w),
+            h = tonumber(meta.h),
+            x = tonumber(meta.x),
+            y = tonumber(meta.y),
+            min_w = mp.get_property_native("width") * options.detect_min_ratio,
+            min_h = mp.get_property_native("height") * options.detect_min_ratio,
+            max_w = mp.get_property_native("width"),
+            max_h = mp.get_property_native("height")
+        }
+    else
+        mp.msg.error("Got empty crop data.")
+        mp.msg.info("You might need to increase detect_seconds.")
+        return
+    end
+
+    apply_crop(meta)
+end
+
+function apply_crop(meta)
+
+    -- Verify if it is necessary to crop.
+    local is_effective = meta.x > 0 or meta.y > 0
+        or meta.w < meta.max_w or meta.h < meta.max_h
+
+    if not is_effective then
+        mp.msg.info("No area detected for cropping.")
+        return
+    end
+
+    -- Verify if it is over cropped.
+    local is_excessive = meta.w < meta.min_w and meta.h < meta.min_h
+
+    if is_excessive then
+        mp.msg.info("The area to be cropped is too large.")
+        mp.msg.info("You might need to decrease detect_min_ratio.")
+        return
+    end
+
+    -- Remove existing crop.
+    remove_filter(labels.crop)
+
+    -- Apply crop.
+    mp.command(
+        string.format("vf pre @%s:lavfi-crop=w=%s:h=%s:x=%s:y=%s",
+            labels.crop, meta.w, meta.h, meta.x, meta.y
         )
-        mp.msg.error(
-            "Does your version of ffmpeg/libav support AVFrame metadata?"
+    )
+end
+
+function on_start()
+
+    -- Clean up at the beginning.
+    cleanup()
+
+    -- If auto is not true, exit.
+    if not options.auto then
+        return
+    end
+
+    -- If it's not a video, exit.
+    if not (typeof_playing() == "video") then
```

Comment:
I just tried cropping album art and images and it doesn't work because there is no crop data since cropdetect ignores the first frame. I don't think there is much to be done about that. This check should probably be run for both modes after all. It could be moved into `detect_crop` which should simplify things.

Reply:
I add it into `on_start()` and `on_toggle()`.
if add into `detect_crop` script will first exit in on_start because of not enough time, and show a warning. this will cause confusion.

Reply:
As with my comment above, not enough time is an accurate description of the still image problem. Better to keep it simple and not duplicate the check.